packages/use-optionally-controlled-state/src/index.tsx

```diff
@@ -1,17 +1,38 @@
 import {useState, useCallback} from 'react';
 import useConstant from 'use-constant';
 
-type Options<Value> =
-  | {
-      controlledValue?: Value;
-      initialValue: Value;
-      onChange?(value: Value): void;
-    }
-  | {
-      controlledValue: Value;
-      initialValue?: Value;
-      onChange?(value: Value): void;
-    };
+// Controlled
+export default function useOptionallyControlledState<Value>({
+  controlledValue,
+  initialValue,
+  onChange
+}: {
+  controlledValue: Value;
+  initialValue?: Value;
+  onChange?(value: Value): void;
+}): [Value, (value: Value) => void];
+
+// Uncontrolled without initial value
+export default function useOptionallyControlledState<Value>({
+  controlledValue,
+  initialValue,
+  onChange
+}: {
+  controlledValue?: Value;
+  initialValue?: Value;
+  onChange?(value: Value): void;
+}): [Value | undefined, (value: Value) => void];
+
+// Uncontrolled with initial value
+export default function useOptionallyControlledState<Value>({
+  controlledValue,
+  initialValue,
+  onChange
+}: {
+  controlledValue?: Value;
+  initialValue: Value;
+  onChange?(value: Value): void;
+}): [Value, (value: Value) => void];
```

Comment:
I had the idea to use function overloading to be able to specify the pairs of arguments and return types together …